history() : Disable Caching instead of Using History ID

danieldresser-ie opened the pull request:
We've already talked through a lot of this.  Might not hurt to have a few more tests if you can think of anything good.  ( 29e86df on the other hand is so specific though that maybe we should omit it ).

Performance all looks good in the tests I have made.

I haven't gone exactly with where we are this morning for naming, though it was all pretty tentative:
* I've gone with disallowCacheRead instead of allowCacheRead, mostly because mayDisallowCacheRead made more sense than having alwaysAllowsCacheRead.  Happy to rename once we think of something better.
* I've currently got mayDisallowCacheRead being called for each cache type - this would allow us without changing client code to implement a per-type flag in the future if it was a useful optimization ( though I haven't done this yet )
* I'm using a new enum instead of Process::type to identify Value vs Hash currently.  I could switch this to being Process::type fairly easily, though the value is hardcoded everywhere it's used - there doesn't seem to be any real advantage to this, and the enum feels easier to understand to me

The final commit where I re-enable writing the cache while reading is disallowed is less tested than the rest, but the tests all pass still, so hopefully it's good.

danieldresser-ie commented:
I've made the switch from "mayDisallow" to "mightDisallow" in place.

I was doing my best to try and implement the "use Process::type to identify type of hash", but the results I was coming up with make so little sense that I think it's better to talk about it first.

HashProcess and ComputeProcess are private to ValuePlug, so when SceneAlgo wants to make a monitor that disallows caching of hashes, it needs a way other than using those to get HashProcess::type.  Rather than have this spread through any usages of "disallow", it would probably make sense to have it centralized somewhere ... does this mean ValuePlug now needs to expose public static accessors for HashProcess::type and ComputeProcess::type, and we then hardcode to them?  Initializing these statics could be tricky, since if we were ever using this heavily, we wouldn't want to require a function call every time they're used.

But perhaps more to the point, even in the hypothetical best case, where we need to use these in ValuePlug itself, we need to check if caching is disabled for HashProcess::type before we decide whether to create a HashProcess, so even there we can't use HashProcess::type directly ... I probably need to understand what the hypothetical benefit in clarity would be before I can implement this well - right now, I'm not seeing anywhere we could actually use a call to the type() function in this context.

danieldresser-ie commented:
Oh, an addendum:  the last commit just comments out a test that was checking that we weren't triggering a history evaluation during update of the Camera transform tool, by checking that GafferScene.SceneAlgo.historyIDContextName() isn't in the context.  Since historyIDContextName() no longer exists, from one respect, it would be safe to delete this line - in a sense, it would always pass now.  But then the test is no longer fufilling its intended function of making sure history() isn't being used ... now that we're intentionally making a history eval more like a normal eval, it's harder to tell if history() has run.  I guess we could add a special monitor that checks if any other monitor has done evals with caching disabled, just for the tests?  Seems a tad convoluted.

johnhaddon commented on include/Gaffer/ThreadState.h:
```diff
@@ -121,7 +122,26 @@ class GAFFER_API ThreadState
 			return m_context;
 		}
 
+		enum class CacheType
+		{
+			Value,
+			Hash
+		};
```
`Process::processType()` is deliberately an `InternedString` rather than an enum to allow it to be an open-ended system : extension modules can introduce their own process types (and have them be monitored) without modifying the core. One example of this is the [various process types](https://github.com/GafferHQ/gaffer/blob/main/src/GafferDispatch/TaskNode.cpp#L91-L123) used by GafferDispatch. To date, none of ThreadState, Monitor or Process have needed any concept of what valid process types might be (or even that ValuePlug exists), and I'd like to keep it that way. That's why I'm advocating for passing `InternedString processType` rather than this new `CacheType` enum.

johnhaddon commented on src/Gaffer/Monitor.cpp:
```diff
@@ -93,7 +101,16 @@ Monitor::Scope::Scope( const MonitorSet &monitors, bool active )
 		}
 	}
+
+	// TODO - the amount of duplication here is pretty ugly in general
```
Agreed. Could we just delegate everything to a private constructor that is passed what to put in `m_monitors`?

johnhaddon commented on src/Gaffer/Monitor.cpp:
```diff
@@ -66,7 +66,15 @@ Monitor::Scope::Scope( const MonitorPtr &monitor, bool active )
+	bool mightDisallowCacheRead = false;
+	for( const auto &m : m_monitors )
+	{
+		mightDisallowCacheRead |= m->mightDisallowCacheRead( ThreadState::CacheType::Value );
+		mightDisallowCacheRead |= m->mightDisallowCacheRead( ThreadState::CacheType::Hash );
```
This is a bit problematic for my "open ended process types" world view. But I think it's probably OK to remove the `type` argument from `mightDisallowCacheRead()`?

johnhaddon commented on python/GafferSceneTest/SceneAlgoTest.py:
```diff
@@ -301,19 +301,31 @@ def testHistory( self ) :
-			attributes["expression"].setExpression(
-				'b = parent["boundQuery"]["bound"]; parent["attributes"]["NameValuePlug"]["value"] = b.size()[0]'
-			)
+		attributes["sourceScene"] = GafferScene.Sphere()
+		attributes["boundQuery"] = GafferScene.BoundQuery()
```
This is pre-existing code, but reading it again I was baffled about why we're parenting these nodes under the StandardAttributes node. I think they can just be local variables instead? Oh, wait, is this some scheme to give the expression a `parent` scope to operate in? Maybe we could avoid all that by just feeding the output from the bound query into a V3f attribute?

johnhaddon commented on python/GafferSceneTest/SceneAlgoTest.py:
```diff
@@ -301,19 +301,31 @@ def testHistory( self ) :
+
+		runTest()
+
+		# Test running the test while everything is already cached still works, and doesn't add any
+		# new entries to the cache
```
> Perhaps unnecessary, since I can't really picture this failing under the new code?

Perhaps, but I think I like it.

johnhaddon commented on include/Gaffer/ValuePlug.h:
```diff
@@ -183,6 +183,8 @@ class GAFFER_API ValuePlug : public Plug
 		/// > Note : Limits are applied on a per-thread basis as and
 		/// > when each thread is used to compute a hash.
 		static void setHashCacheSizeLimit( size_t maxEntriesPerThread );
+		/// Returns the total number of entries in both global and per-thread hash caches
```
We should add this to the API section of Changes.md.

johnhaddon commented on include/Gaffer/ThreadState.h:
```diff
@@ -116,6 +116,11 @@ class GAFFER_API ThreadState
 
 		static const ThreadState &current();
 
+		const Context *getContext() const
```
Since there's no equivalent `setContext()`, this would just be `context()`.

johnhaddon commented on include/GafferScene/SceneAlgo.h:
```diff
@@ -215,12 +215,6 @@ GAFFERSCENE_API SceneProcessor *objectTweaks( const ScenePlug *scene, const Scen
-/// to disable the effects of the hash cache, so that the full upstream process can be examined.
-/// May be removed from a context to reenable the cache for expensive hash operations that are
-/// known to be irrelevant to the history.
-GAFFERSCENE_API IECore::InternedString historyIDContextName();
```
We should note the removal in the Breaking Changes section of Changes.md.

johnhaddon commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -489,7 +505,6 @@ SceneAlgo::History::Ptr historyWalk( const CapturedProcess *process, InternedStr
 		if( scene && plug == scene->getChild( scenePlugChildName ) )
 		{
 			ContextPtr cleanContext = new Context( *process->context );
-			cleanContext->remove( SceneAlgo::historyIDContextName() );
```
It looks like `cleanContext` is unecessary now, and we can just reference `process->context` directly.

johnhaddon commented on src/Gaffer/ValuePlug.cpp:
```diff
@@ -239,11 +239,17 @@ class ValuePlug::HashProcess : public Process
+			else if( threadState.disallowCacheRead( plug, ThreadState::CacheType::Hash ) )
+			{
+				HashProcess process( processKey );
+				g_globalCache.set( processKey, process.m_result, 1 );
```
Depending on the cache policy, this might not be the right cache to set it in. We only ever perform lookups in the global cache for the `TaskCollaboration` and `TaskIsolation` policies, and we rely on very low eviction rates from the global cache. So for policies that use the local cache alone, we should set into the local cache.

danieldresser-ie commented on include/Gaffer/Monitor.h:
```diff
@@ -83,6 +83,21 @@ class GAFFER_API Monitor : public IECore::RefCounted
+		/// Must return true if disallowCacheRead will ever return true from this Monitor
+		/// ( for the given type )
+		virtual bool mightDisallowCacheRead( ThreadState::CacheType type )
+		{
+			return false;
+		}
```
In this case, to my view, looking at the header as "information that will be useful to implementor", the fact that these functions default to returning false seems like a useful thing to advertise.  But since they're virtual, there's no performance benefit.  Moved to cpp under protest.

danieldresser-ie commented:
Tidying up all these loose ends is taking a bit longer than expected.

The important thing was the performance questions.  I was able to reproduce the slowdown with some reliability by reducing the iterations in repeatGetValue from 5000000 to 500000, and running the test with a repeat of 100 - this gives me a 2% slowdown as well.  I realized that I'd been thinking about the CachePolicy::Standard case, but what we're actually using by default currently is Legacy.  The slowdown is because the Legacy case doesn't use the ThreadState that I get from ThreadState::current(), so that call is pure overhead.  Using Standard gives a 5% improvement over Legacy, and that time isn't affected by my changes, because the ThreadState is needed anyway.

So I think these changes are probably OK, with the caveat that we should really look in general at ways to avoiding needing to call ThreadState::current()?

Other than that, I've addressed most feedback, though I should probably still look over the changes again.  Remaining on the checklist:  add changelog entries, figure out what to do with CameraToolTest.testNoUnecessaryHistoryCalls.  You had suggested clearing the cache, and then testing that it's still empty - this doesn't work, there is stuff in the cache at the end.  I haven't confirmed why yet, but although we don't want computes triggered during the camera move, perhaps there are computes legitimately triggered when the move finishes?  Haven't really investigated yet.

danieldresser-ie commented:
I went down way too deep a rabbit hole of trying to understand testNoUnecessaryHistoryCalls, and still don't quite understand it.  I was trying to prepopulate the caches, so that viewportCameraChanged wouldn't trigger computes, as you suggested.  But as far as I can tell, since the test is actually moving the camera, the camera transform gets edited, at which point, the compute will happen, and there's no good way to tell the difference.

I had pretty well convinced myself that the right answer was MonitorMonitor, which would monitor which monitors are active during hashes and computes, but then I realized Monitors don't even have a typeId to identify themselves by, so I gave up completely and just put a comment in testNoUnecessaryHistoryCalls.

I have added in the Changes.md entries now.

johnhaddon reviewed (changes requested):
> I went down way too deep a rabbit hole of trying to understand testNoUnecessaryHistoryCalls, and still don't quite understand it. I was trying to prepopulate the caches, so that viewportCameraChanged wouldn't trigger computes, as you suggested. But as far as I can tell, since the test is actually moving the camera, the camera transform gets edited, at which point, the compute will happen, and there's no good way to tell the difference.

This seems to do the job :arrow_down:. The test passes, and if I put a bogus `SceneAlgo::history()` call into the CameraTool, the test fails. I think you were probably being thwarted by the unnecessary forcing of an update _after_ the edit, which was triggering a compute.

```
with Gaffer.PerformanceMonitor() as pm :

	view.viewportGadget().setCameraTransform(
		imath.M44f().translate( imath.V3f( 1, 2, 3 ) )
	)

	# The change to the viewport camera should have been mirrored
	# onto the Camera node by the CameraTool.
	self.assertEqual(
		script["camera"]["transform"]["translate"].getValue(),
		imath.V3f( 1, 2, 3 )
	)

# We do not want the CameraTool to have performed a `SceneAlgo::history()`
# query during the edit, as they can be expensive and aren't suitable
# for repeated use during drags etc. Any use of `history()` will show
# up in our PerformanceMonitor.

self.assertEqual( len( pm.allStatistics() ), 0 )
```

danieldresser-ie commented:
Addressed comments.  It's a bit concerning that we don't have enough test coverage to notice that I'd reversed that conditional for where we cache, but that doesn't seem like an easy thing to test.

johnhaddon commented:
> It's a bit concerning that we don't have enough test coverage to notice that I'd reversed that conditional for where we cache, but that doesn't seem like an easy thing to test.

I think this is testable actually :

- Start with an empty cache (all tests start this way anyway)
- Do a `history()` query
- Call `attributes()`, inside a PerformanceMonitor
- Check that the PerformanceMonitor didn't see any processes

The hash for the attributes is meant to end up in the local cache, but if it ends up in the global cache we won't find it and will trigger a compute.

johnhaddon commented:
>> It's a bit concerning that we don't have enough test coverage to notice that I'd reversed that conditional for where we cache, but that doesn't seem like an easy thing to test.

> I think this is testable actually

I've merged anyway, since it'd be good to have this over the line before I head off on holiday. Might be good to open another PR to add that extra bit of test coverage though.
